From 918da17b0be2286382772ee38d53311b0be146b4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20=C3=85dahl?= <jadahl@gmail.com>
Date: Mon, 1 Jun 2026 22:48:44 +0200
Subject: [PATCH] cursor/wayland: Don't hold a ref on the surface

The surface owns a reference to the Wayland cursor instance tied to it,
so if the cursor itself has a reference back to the surface, we end up
with a circular dependency, causing a memory leak. Fix this by making
the Wayland cursor instance only have a weak pointer to the surface, and
handle the surface disappearing gracefully.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/5097>
(cherry picked from commit e8d998b52af31347d2ce83c39ae7eaa4c4d28586)
---
 src/wayland/meta-cursor-wayland.c | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/src/wayland/meta-cursor-wayland.c b/src/wayland/meta-cursor-wayland.c
index c345eb63741..16db71a8f9b 100644
--- a/src/wayland/meta-cursor-wayland.c
+++ b/src/wayland/meta-cursor-wayland.c
@@ -85,11 +85,15 @@ meta_cursor_wayland_prepare_at (ClutterCursor *cursor,
   MetaCursorTracker *cursor_tracker = cursor_wayland->cursor_tracker;
   MetaBackend *backend =
     meta_cursor_tracker_get_backend (cursor_tracker);
-  MetaWaylandSurface *surface = cursor_wayland->surface;
   MetaMonitorManager *monitor_manager =
     meta_backend_get_monitor_manager (backend);
+  MetaWaylandSurface *surface;
   MetaLogicalMonitor *logical_monitor;
 
+  surface = cursor_wayland->surface;
+  if (!surface)
+    return;
+
   logical_monitor =
     meta_monitor_manager_get_logical_monitor_at (monitor_manager, x, y);
   if (logical_monitor)
@@ -182,7 +186,7 @@ meta_cursor_wayland_finalize (GObject *object)
   MetaCursorWayland *cursor_wayland = META_CURSOR_WAYLAND (object);
 
   g_clear_object (&cursor_wayland->texture);
-  g_clear_object (&cursor_wayland->surface);
+  g_clear_weak_pointer (&cursor_wayland->surface);
 
   G_OBJECT_CLASS (meta_cursor_wayland_parent_class)->finalize (object);
 }
@@ -230,7 +234,7 @@ meta_cursor_wayland_new (MetaWaylandSurface *surface,
   cursor_wayland = g_object_new (META_TYPE_CURSOR_WAYLAND,
                                  "color-state", color_state,
                                  NULL);
-  g_set_object (&cursor_wayland->surface, surface);
+  g_set_weak_pointer (&cursor_wayland->surface, surface);
   cursor_wayland->cursor_tracker = cursor_tracker;
 
   return cursor_wayland;
@@ -239,7 +243,10 @@ meta_cursor_wayland_new (MetaWaylandSurface *surface,
 MetaWaylandBuffer *
 meta_cursor_wayland_get_buffer (MetaCursorWayland *cursor_wayland)
 {
-  return meta_wayland_surface_get_buffer (cursor_wayland->surface);
+  if (cursor_wayland->surface)
+    return meta_wayland_surface_get_buffer (cursor_wayland->surface);
+  else
+    return NULL;
 }
 
 static void
-- 
GitLab

